specs/DragStarting.md

```diff
@@ -0,0 +1,196 @@
+DragStarting
+===
+
+# Background
+The WebView2 team has been asked to provide a way to override the default drag drop behavior when running in visual hosting mode. This event allows you to know when a drag is initiated in WebView2 and provides the state necessary to override the default WebView2 drag operation with your own logic.
+
+# Examples
+## DragStarting
+Users can use `add_DragStarting` on the CompositionController to add an event handler
+that is invoked when drag is starting. They can use the the event args to start their own
+drag. Notably the `Deferral` can be used to execute any async drag logic and call back into
+the WebView at a later time. The `Handled` property lets the WebView2 know whether to
+exercise its own drag logic or not.
+
+```C++
+CHECK_FAILURE(m_compControllerStaging->add_DragStarting(
+  Callback<ICoreWebView2StagingDragStartingEventHandler>(
+      [this](ICoreWebView2CompositionController* sender, ICoreWebView2StagingDragStartingEventArgs* args)
+      {
+          if (m_dragOverrideMode != DragOverrideMode::OVERRIDE)
+          {
+              // If the event is marked handled, WebView2 will not execute its drag logic.
+              args->put_Handled(m_dragOverrideMode == DragOverrideMode::NOOP);
+              return S_OK;
+          }
+
+          COREWEBVIEW2_DRAG_EFFECTS allowedEffects;
+          POINT dragPosition;
+          wil::com_ptr<IDataObject> dragData;
+          wil::com_ptr<ICoreWebView2Deferral> deferral;
+
+          args->get_DragAllowedOperations(&allowedEffects);
+          args->get_DragPosition(&dragPosition);
+          args->get_DragData(&dragData);
+          args->GetDeferral(&deferral);
+
+          HRESULT hr = S_OK;
+          if (!m_oleInitialized)
+          {
+              hr = OleInitialize(nullptr);
+              if (SUCCEEDED(hr))
+              {
+                  m_oleInitialized = true;
+              }
+          }
+
+          if (!m_dropSource)
+          {
+              m_dropSource = Make<ScenarioDragDropOverrideDropSource>();
+          }
+
+          if (SUCCEEDED(hr))
+          {
+              DWORD effect;
+              DWORD okEffects = DROPEFFECT_NONE;
+              if (allowedEffects & COREWEBVIEW2_DRAG_EFFECTS_COPY)
+              {
+                  okEffects |= DROPEFFECT_COPY;
+              }
+              if (allowedEffects & COREWEBVIEW2_DRAG_EFFECTS_MOVE)
+              {
+                  okEffects |= DROPEFFECT_MOVE;
+              }
+              if (allowedEffects & COREWEBVIEW2_DRAG_EFFECTS_LINK)
+              {
+                  okEffects |= DROPEFFECT_LINK;
+              }
+
+              hr = DoDragDrop(dragData.get(), m_dropSource.get(), okEffects, &effect);
+          }
+
+          deferral->Complete();
```

Comment:
You have a deferral but this all seems like sync code. Is DoDragDrop an OS API that runs a message loop and so we can have reentering code and that's why we need the deferral? Or is the deferral not necessary?